src/tpm2_createpolicy.c

```diff
@@ -0,0 +1,330 @@
+
+#define INIT_SIMPLE_TPM2B_SIZE( type ) (type).t.size = sizeof( type ) - 2;
+
+#define SET_PCR_SELECT_BIT( pcrSelection, pcr ) \
+ (pcrSelection).pcrSelect[( (pcr)/8 )] |= ( 1 << ( (pcr) % 8) );
+
+TPM_RC build_pcr_policy( createpolicypcr_ctx *ctx) {
+
+    TPM_RC rval = TPM_RC_SUCCESS;
+    TPML_PCR_SELECTION pcrs;
+    TPML_DIGEST pcrValues;
+    UINT32 pcrUpdateCounter;
+    TPM2B_DIGEST pcrDigest;
+    TPML_PCR_SELECTION pcrSelectionOut;
+    pcrs.count = 1;
+    pcrs.pcrSelections[0].hash = ctx->hash_alg;
+    pcrs.pcrSelections[0].sizeofSelect = 3; 
+    pcrs.pcrSelections[0].pcrSelect[0] = 0; 
+    pcrs.pcrSelections[0].pcrSelect[1] = 0; 
+    pcrs.pcrSelections[0].pcrSelect[2] = 0; 
+    SET_PCR_SELECT_BIT( pcrs.pcrSelections[0], ctx->pcr_index); 
+
+    FILE *fp;
+    struct stat st;
+    int errno =0, i=0, sz=0;
+    if(ctx->raw_pcr_flag) {
+        //digests[count{2:}]{:8}
+        pcrValues.count = pcrs.count;
+
+        switch(ctx->hash_alg) {
+            case TPM_ALG_SHA1: 
+                pcrValues.digests[0].t.size = SHA1_DIGEST_SIZE;
+                break;
+            case TPM_ALG_SHA256:
+                pcrValues.digests[0].t.size = SHA256_DIGEST_SIZE;
+                break;
+            case TPM_ALG_SHA384:
+                pcrValues.digests[0].t.size = SHA384_DIGEST_SIZE;
+                break;            
+            case TPM_ALG_SHA512:
+                pcrValues.digests[0].t.size = SHA512_DIGEST_SIZE;
+                break;
+            case TPM_ALG_SM3_256:
+                pcrValues.digests[0].t.size = SM3_256_DIGEST_SIZE;
+                break;
+            default:
+                pcrValues.digests[0].t.size = SHA1_DIGEST_SIZE;
+                break;
+        }
+        if (stat(ctx->raw_pcr_file, &st) == 0) {
+            if(st.st_size != pcrValues.digests[0].t.size) {
+                printf("pcr input file has wrong size.\n");
+                return -1;
+            }
+        }
+        else {
+            fprintf(stderr, "Cannot determine size of %s: %s\n",
+            ctx->raw_pcr_file, strerror(errno));
+            return -1;
+        }
+        fp = fopen(ctx->raw_pcr_file, "rb");
+        for(i=0; i<pcrValues.digests[0].t.size; i++) {
+            sz = fread(&pcrValues.digests[0].t.buffer[i],1,1,fp);
+        }
+        fclose(fp);
+    } 
+    else {
+        // Read PCRs
+        rval = Tss2_Sys_PCR_Read( ctx->sapi_context, 0, &pcrs, &pcrUpdateCounter, 
+            &pcrSelectionOut, &pcrValues, 0 );
+        if( rval!=TPM_RC_SUCCESS ){
+            return rval;
+        }    
+    }
+    
+
+    // Calculate digest( with authhash alg) of pcrvalues in variable pcrDigest
+    pcrDigest.t.size = 0;
+    INIT_SIMPLE_TPM2B_SIZE( pcrDigest );
+    rval = tpm_hash_sequence( ctx->sapi_context, ctx->policy_session->authHash,
+        pcrValues.count, &pcrValues.digests[0], &pcrDigest);
+    if( rval!=TPM_RC_SUCCESS ){
+        return rval;
+    }
+    rval = Tss2_Sys_PolicyPCR( ctx->sapi_context, 
```

Comment:
just return Tss2_Sys_PolicyPCR() here, the branches below don't change anything, plus you pick up the tail call optimization.